File: problem_solving/linkedlist/py/430_flatten_multilevel_double_linklist.py
'''
flatten_multilevel_double_linklist
https://leetcode.com/problems/flatten-a-multilevel-doubly-linked-list/

look for node where child is not null
    apply the recursive to the child

    set up links between current and the child.. next and prev
    set up links between saved next and the child tail.

O(N)
O(1)
'''

class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child


from linkedlist import ListNode



def flatten(head: "Node") -> "Node":

    def flatten_dfs(node: "Node"):
        current = node
        tail = node
        while current:
            next_node = current.next
            if current.child:
                child_tail = flatten_dfs(current.child)
                current.next = current.child
                current.child.prev = current
                current.child = None

                if next_node:
                    child_tail.next = next_node
                    next_node.prev = child_tail

                tail = child_tail
                current = next_node
            else:
                tail = current
                current = current.next
        return tail        
        
    if not head:
        return None
    flatten_dfs(head)
    return head


    

File: problem_solving/linkedlist/cpp/430_flatten_multilevel_double_linklist.cpp
/*
'''
flatten_multilevel_double_linklist
https://leetcode.com/problems/flatten-a-multilevel-doubly-linked-list/

use of stack to store the next when child node is encountered.
look for node where child is not null
    apply the recursive to the child

    set up links between current and the child.. next and prev
    set up links between saved next and the child tail.

O(N)
O(1)
'''
*/

#include<iostream>
#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"
using namespace std;
#include<algorithm>
#include<stack>


// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};


Node* flatten(Node* head) {
    if(!head) {
        return nullptr;
    }
    Node* node = head;
    std::stack<Node*> st;

    while(node || !st.empty()) {
        if(node && node->child) {
            //if there is a child node
            if(node->next) {
                st.push(node->next); //push to stack.
            }
            node->next = node->child;//go to the child node.
            node->child->prev = node;
            node->child = nullptr;
        } 
        if(!node->next && !st.empty()) {
            //we are at the end of the list,,. check stack.
            auto next_node = st.top();
            st.pop();
            if(next_node) {
                node->next = next_node;
                next_node->prev = node;
            }
        }
        node = node->next;
    }
    return head;
}

int main() {
    return 0;
}


    


